packages/gatsby-source-contentful/src/types.js

```diff
@@ -0,0 +1,406 @@
+        // but I can't create actual "children" field which
+        // is union combining all children node types
+      },
+      interfaces: [`Node`],
+      extensions: {
+        infer: false,
+      },
+    })
+
+    context.gqlTypes.push(textGQLTypeConfig)
+
+    // use that type
+    return {
+      type: textTypeName,
+      extensions: {
+        link: {
+          from: `${context.fieldName}___NODE`,
+        },
+      },
+    }
+  },
+
+  Boolean: () => `Boolean`,
+
+  // Reference and Media fields
+  Link: (field, context) => {
+    if (field.linkType === `Asset`) {
+      // Media field
+      const type = {
+        type: `ContentfulAsset`,
+        extensions: {
+          link: {
+            from: `${context.fieldName}___NODE`,
+          },
+        },
+      }
+
+      return type
+    } else if (field.linkType === `Entry`) {
+      // Reference field
+      let contentTypes = null
+
+      // If there is validations field:
+      if (field.validations) {
+        // lok for validations that limit available content types
+        field.validations.some(validation => {
+          if (validation.linkContentType) {
+            contentTypes = context.contentTypeItems.filter(contentType =>
+              validation.linkContentType.includes(contentType.sys.contentful_id)
+            )
+            return true
+          }
+          return false
+        })
+      }
+
+      if (!contentTypes) {
+        contentTypes = context.contentTypeItems
+      }
+
+      // add backreferences
+      contentTypes.forEach(contentType => {
+        context.addBackReference({
+          referencingType: context.typeName,
+          referencedType: contentType.name,
+        })
+      })
+
+      if (contentTypes.length > 1) {
+        // this is union field
+        const usedTypesNames = contentTypes.map(contentType =>
+          makeTypeName(contentType.name)
+        )
+
+        // this is re-implementation of union type naming function from gatsby
+        const UnionTypeName = usedTypesNames.sort().join(``) + `Union`
+
+        let unionType = context.gqlTypes.find(
+          gqlType =>
+            gqlType.kind === `UNION` && gqlType.config.name === UnionTypeName
+        )
+
+        // create union type, if it doesn't exist yet
+        if (!unionType) {
+          unionType = context.schema.buildUnionType({
+            name: UnionTypeName,
+            types: usedTypesNames,
+            resolveType: node => node.internal.type,
+          })
+
+          context.gqlTypes.push(unionType)
+        }
+
+        // use that type
+        return {
+          type: UnionTypeName,
+          extensions: {
+            link: {
+              from: `${context.fieldName}___NODE`,
+            },
+          },
+        }
+      } else if (contentTypes.length === 1) {
+        // this is single type - not sure if this should be special case
+        // (using regular object type)
+        // or should create union with single type?
```

Comment:
Probably should create a Union with single type, to force a spread inside query. That would mean that if more content types are added, query is still backwards compatible.

Reply:
My biggest worry here is that unions and inline fragments needed to properly query them seem to be common source of confusion for people using graphql for the first time (graphiql doesn't help much - by default people think they can query jsut for `__typename` - but documenting how to query union fields should work here)

Reply:
graphiql explorer actually shows them which is really nice https://github.com/gatsbyjs/gatsby/issues/11602

Reply:
Yeah, graphiql explorer would indeed nullify my worry here:
![Kapture 2019-05-22 at 21 53 14](https://user-images.githubusercontent.com/419821/58204542-c1591180-7cdc-11e9-881a-59366124f8dc.gif)

Reply:
https://github.com/gatsbyjs/gatsby/pull/14280 is close to be finished, so I will go ahead and change to use Union here so references will always create unions

Reply:
I started to think if we should still do 1-type unions here. Was doing more testing on more real project and queries really looks messy with this - see https://github.com/gatsbyjs/gatsby/pull/12816#issuecomment-515762555